src/ctap2/pin.rs

```diff
@@ -1,8 +1,241 @@
+    One,
+}
+
+#[derive(Debug)]
+pub struct PinProtocolState {
+    key_agreement_key: KeyId,
+    // only used to delete the old shared secret from VFS when generating a new one.  ideally, the
+    // SharedSecret struct would clean up after itself.
+    shared_secret: Option<KeyId>,
+
+    // for protocol version 1
+    pin_token_v1: KeyId,
+}
+
+impl PinProtocolState {
+    // in spec: initialize(...)
+    pub fn new<T: TrussedRequirements>(trussed: &mut T) -> Self {
+        Self {
+            key_agreement_key: generate_key_agreement_key(trussed),
+            shared_secret: None,
+            pin_token_v1: generate_pin_token(trussed),
+        }
+    }
+
+    pub fn reset<T: TrussedRequirements>(self, trussed: &mut T) {
+        syscall!(trussed.delete(self.pin_token_v1));
+        syscall!(trussed.delete(self.key_agreement_key));
+        if let Some(shared_secret) = self.shared_secret {
+            syscall!(trussed.delete(shared_secret));
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct PinProtocol<'a, T: TrussedRequirements> {
+    trussed: &'a mut T,
+    state: &'a mut PinProtocolState,
+    version: PinProtocolVersion,
+}
+
+impl<'a, T: TrussedRequirements> PinProtocol<'a, T> {
+    pub fn new(
+        trussed: &'a mut T,
+        state: &'a mut PinProtocolState,
+        version: PinProtocolVersion,
+    ) -> Self {
+        Self {
+            trussed,
+            state,
+            version,
+        }
+    }
+
+    fn pin_token(&self) -> KeyId {
+        match self.version {
+            PinProtocolVersion::One => self.state.pin_token_v1,
+        }
+    }
+
+    fn set_pin_token(&mut self, pin_token: KeyId) {
+        match self.version {
+            PinProtocolVersion::One => self.state.pin_token_v1 = pin_token,
+        }
+    }
+
+    pub fn regenerate(&mut self) {
+        syscall!(self.trussed.delete(self.state.key_agreement_key));
+        if let Some(shared_secret) = self.state.shared_secret.take() {
+            syscall!(self.trussed.delete(shared_secret));
+        }
+        self.state.key_agreement_key = generate_key_agreement_key(self.trussed);
+    }
+
+    // in spec: resetPinUvAuthToken()
+    pub fn reset_pin_token(&mut self) {
+        syscall!(self.trussed.delete(self.pin_token()));
+        let pin_token = generate_pin_token(self.trussed);
+        self.set_pin_token(pin_token);
+    }
+
+    // in spec: getPublicKey
+    #[must_use]
+    pub fn key_agreement_key(&mut self) -> EcdhEsHkdf256PublicKey {
+        let public_key = syscall!(self
+            .trussed
+            .derive_p256_public_key(self.state.key_agreement_key, Location::Volatile))
+        .key;
+        let serialized_cose_key = syscall!(self.trussed.serialize_key(
+            Mechanism::P256,
+            public_key,
+            KeySerialization::EcdhEsHkdf256
+        ))
+        .serialized_key;
+        let cose_key = cbor_deserialize(&serialized_cose_key).unwrap();
+        syscall!(self.trussed.delete(public_key));
```

Comment:
Not really, but I don’t see a good way around it currently.  But Trussed should definitely provide helper methods for deserialization to avoid incompatibilities.

I think `cosey` is just a copy of `ctap_types::cose` (or vice versa).  They should definitely be merged at some point.